File: src/main/scala/com/github/tetrisanalyzer/move/ValidMoves.scala
package com.github.tetrisanalyzer.move

import com.github.tetrisanalyzer.piecemove.PieceMove
import com.github.tetrisanalyzer.board.Board

object ValidMoves {
  def apply(board: Board) = {
    new ValidMoves(board)
  }
}

class ValidMoves(board: Board) {
  private var validMoves = List.empty[PieceMove]
  private val visitedMoves = Array.fill(board.height, board.width) { 0 }

  private def markAsVisited(move: Move) { visitedMoves(move.y)(move.x) |= 1 << move.rotation }
  private def isUnvisited(move: Move): Boolean = {
    try {
      (visitedMoves(move.y)(move.x) & (1 << move.rotation)) == 0
    } catch {
      case e: IndexOutOfBoundsException => false
    }
  }

  /**
   * The incoming pieceMove is at the starting position of the board, and
   * links to all possible moves on an empty board. Because the board is probably not empty,
   * every every possible move needs to be checked (and returned by this method).
   */
  def pieceMoves(pieceMove: PieceMove): List[PieceMove] = {
    if (isUnvisited(pieceMove.move)) {
      markAsVisited(pieceMove.move)

      pieceMove.freeAsideAndRotateMoves.foreach(move => pieceMoves(move))

      if (pieceMove.canMoveDown)
        pieceMoves(pieceMove.down)
      else
        validMoves = pieceMove :: validMoves
    }
    validMoves
  }
}

File: langs/java/src/main/java/com/github/tetrisanalyzer/move/ValidMoves.java
package com.github.tetrisanalyzer.move;

import com.github.tetrisanalyzer.board.Board;
import com.github.tetrisanalyzer.piecemove.PieceMove;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ValidMoves {
    private final List<PieceMove> validMoves = new ArrayList<PieceMove>();
    private final int[][] visitedMoves;

    public ValidMoves(Board board) {
        visitedMoves = getEmptyVisitedMoves(board.height, board.width);
    }

    private int[][] getEmptyVisitedMoves(int height, int width) {
        int[][] visitedMoves = new int[height][];
        for (int h=0; h<height; h++) {
            visitedMoves[h] = new int[width];
            Arrays.fill(visitedMoves[h], 0);
        }
        return visitedMoves;
    }

    private void markAsVisited(Move move) {
        visitedMoves[move.y][move.x] |= 1 << move.rotation;
    }

    private boolean isUnvisited(Move move) {
        try {
            return (visitedMoves[move.y][move.x] & (1 << move.rotation)) == 0;
        } catch(IndexOutOfBoundsException e) {
            return false;
        }
    }

    /**
     * The incoming pieceMove is at the starting position of the board, and
     * links to all possible moves on an empty board. Because the board is probably not empty,
     * every possible move needs to be checked (and returned by this method).
     */
    public List<PieceMove> getPieceMoves(PieceMove pieceMove) {
        if (isUnvisited(pieceMove.move) && pieceMove.isFree()) {
            markAsVisited(pieceMove.move);

            for (PieceMove move: pieceMove.asideAndRotate) {
                getPieceMoves(move);
            }
            if (pieceMove.canMoveDown()) {
                getPieceMoves(pieceMove.down);
            } else {
                validMoves.add(pieceMove);
            }
        }
        return validMoves;
    }
}
